engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java

```diff
@@ -203,6 +205,18 @@ boolean hasFunctionalDispatch() {
     return true;
   }
 
+  @ExportMessage
+  @CompilerDirectives.TruffleBoundary
+  boolean isDate() {
+    return InteropLibrary.getUncached().isDate(fields[0]);
```

Comment:
Delegating to `fields[0]` is a bit wild - I guess I should first learn what's the meaning of `fields` array.

Reply:
Yeah this seems too permissive. Our atoms are algebraic datatypes, `fields` is the array of all their members. So if I have `type User (created_at : Date) (name : Text) (email : Text)`, this code makes it a date now.
I see two options here:
1. We do it through an interface, i.e. this delegates to an `is_date` method (defined in Enso), `asDate` delegates to `as_date`. This is very nice because it makes things extensible, but there may be accidental clashes. Maybe something like `is_polyglot_date`, to make the Enso-side method less ambiguous in what it does.
2. We verify the constructor is one of the few we know about.

I'd lean towards 1., as I can see value in first-class interop in this language and don't want to presume "what we've built into stdlib is all anyone will ever need"

Reply:
Thanks for the review. Yes delegation to `fields[0]` was just an attempt which would really go bad with the `type User (created_at : Date)....`. Thanks for the two options.

In contrary to you, I lean towards 2. Certainly as a first shot implementation. I assume it is going to be easier - also, with respect to the _opposite direction of the conversion_...

Reply:
Only converting the right atom since [48bc4f1](https://github.com/enso-org/enso/pull/3374/commits/48bc4f16f8f85a14ff3b80adf92575ca5365d8cf)